protocol/binary/stream_reader.go

```diff
@@ -67,11 +68,33 @@ type StreamReader struct {
 	buffer [8]byte
 }
 
-// NewStreamReader returns a new StreamReader.
-func NewStreamReader(r io.Reader) StreamReader {
+// newStreamReader returns a new StreamReader.
+func newStreamReader(r io.Reader) StreamReader {
 	return StreamReader{reader: r}
 }
 
+var streamReaderPool = sync.Pool{
+	New: func() interface{} {
+		return &StreamReader{}
+	}}
+
+// BorrowStreamReader fetches a StreamReader from the system that will write
+// its output to the given io.Reader.
+//
+// This StreamReader must be returned back using ReturnStreamReader.
+func BorrowStreamReader(r io.Reader) *StreamReader {
```

Comment:
Note to whoever takes over this PR: 
Now that there's a Close method on the interface, we might be able to hide the Borrow-Return as an implementation detail. We can rename BorrowStreamReader to `NewStreamReader` and hide `ReturnStreamReader` (since Close will call it).